File: skills/math-expression-parser/benchmarks/python/bench.py
"""
Math Expression Parser benchmark — Python

Measures wall-clock time for the evaluate node.
Outputs NDJSON, one line per node.
"""

import json
import sys
import time

sys.path.insert(0, "../../../../research/experiments/mathexpr-skill-python")
from mathexpr import calc


def benchmark_node(name, iterations, warmup, fn, check):
    correct = check()
    if not correct:
        print(f"FAIL: {name} correctness check failed", file=sys.stderr)
        sys.exit(1)

    for _ in range(warmup):
        fn()

    times = []
    for _ in range(iterations):
        t0 = time.perf_counter_ns()
        fn()
        elapsed_ms = (time.perf_counter_ns() - t0) / 1_000_000
        times.append(elapsed_ms)

    times.sort()
    n = len(times)

    result = {
        "node": name,
        "language": "python",
        "iterations": iterations,
        "warmup": warmup,
        "wall_clock_ms": {
            "min": times[0],
            "median": times[n // 2],
            "p95": times[int(n * 0.95)],
            "max": times[-1],
        },
        "correctness": correct,
    }
    print(json.dumps(result))


# ---------------------------------------------------------------------------
# Workloads
# ---------------------------------------------------------------------------

EXPR = "((2.5 + 3.1) * 4 - 1) ** 2 / 5 + 7 % 3"
EXPECTED = 92.592

benchmark_node(
    "evaluate",
    10000,
    1000,
    lambda: calc(EXPR),
    lambda: abs(calc(EXPR) - EXPECTED) < 1e-6,
)

File: skills/math-expression-parser/benchmarks/go/bench.go
// Math Expression Parser benchmark — Go
//
// Measures wall-clock time for the evaluate node.
// Outputs NDJSON, one line per node.
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
	"time"

	"mathexpr"
)

type wallClock struct {
	Min    float64 `json:"min"`
	Median float64 `json:"median"`
	P95    float64 `json:"p95"`
	Max    float64 `json:"max"`
}

type benchResult struct {
	Node        string    `json:"node"`
	Language    string    `json:"language"`
	Iterations  int       `json:"iterations"`
	Warmup      int       `json:"warmup"`
	WallClockMs wallClock `json:"wall_clock_ms"`
	Correctness bool      `json:"correctness"`
}

func benchmarkNode(name string, iterations, warmup int, fn func(), check func() bool) {
	correct := check()
	if !correct {
		fmt.Fprintf(os.Stderr, "FAIL: %s correctness check failed\n", name)
		os.Exit(1)
	}

	for i := 0; i < warmup; i++ {
		fn()
	}

	times := make([]float64, iterations)
	for i := 0; i < iterations; i++ {
		t0 := time.Now()
		fn()
		times[i] = float64(time.Since(t0).Nanoseconds()) / 1_000_000
	}

	sort.Float64s(times)

	result := benchResult{
		Node:       name,
		Language:   "go",
		Iterations: iterations,
		Warmup:     warmup,
		WallClockMs: wallClock{
			Min:    times[0],
			Median: times[len(times)/2],
			P95:    times[int(float64(len(times))*0.95)],
			Max:    times[len(times)-1],
		},
		Correctness: correct,
	}

	b, _ := json.Marshal(result)
	fmt.Println(string(b))
}

func main() {
	expr := "((2.5 + 3.1) * 4 - 1) ** 2 / 5 + 7 % 3"
	expected := 92.592

	benchmarkNode(
		"evaluate",
		10000,
		1000,
		func() {
			mathexpr.Calc(expr)
		},
		func() bool {
			r, err := mathexpr.Calc(expr)
			if err != nil {
				return false
			}
			return math.Abs(r-expected) < 1e-6
		},
	)
}
